R/data-pipeline/2-data_aggregation.R

```diff
@@ -0,0 +1,108 @@
+# Functions to aggregate data at each intersection of stratifying variables
+#
+# Example
+# base_data <- readRDS(here("data", "processed", "df_base.RDS"))
+# fup_data <- readRDS(here("data", "processed", "df_fup.RDS"))
+# data_id <- clean_data(base_data, fup_data)
+
+pacman::p_load(dplyr, tidyr, purrr)
+
+# summarise by any given strata ------------------------------------
+summarise_ids_by <- function(data, group_cols) {
+
+  # summarise participants per group -----
+  df_participants <- data |>
+    group_by(across(all_of(group_cols))) |>
+    summarise(
+      # participants ---
+      participant_n = length(unique(id)),
+      participant_days_n = n(),
+      # data quality ---
+      ## % records missing weight observations
+      obs_recorded = sum(!is.na(weight)),
+      obs_missing = sum(is.na(weight)),
+      obs_anomalous = sum(!include_observation),
+      obs_excluded_percent = (obs_missing + obs_anomalous) /
+        participant_days_n * 100) |>
+    ungroup()
+
+  # |>
+  #   pivot_longer(cols = -group_cols, names_to = "variable") |>
+  #   mutate(stat = ifelse(grepl("_percent", variable),
+  #                        "percent", "count"))
+
+  # summarise observed metrics -----
+  ## drop anomalous observations
+  data <- data |>
+    filter(include_observation)
+
+  # averages
+  df_centraltendency <- data |>
+    group_by(across(all_of(group_cols))) |>
+    summarise(
+      across(c("weight",
+               "bmi",
+               "percent_change_firstmeasurement",
+               "percent_change_prewar"),
+             .fns = list(
+               mean = ~ mean(., na.rm = TRUE),
+               median = ~ median(., na.rm = TRUE),
+               q1 = ~ quantile(., probs = 0.25, na.rm = TRUE),
+               q3 = ~ quantile(., probs = 0.75, na.rm = TRUE)),
+             .names = "{.col}.{.fn}")
+    ) |>
+    pivot_longer(cols = -group_cols) %>%
+    separate(name, into = c("variable", "stat"), sep = "\\.")
+
+  # proportions
```

Comment:
included BMI category percentages in this function so it is kept together with other summarising code and produces a single output